Automattic/jetpack
# AI Extension: handle errors from extended block actions

Fixes #31382.

## Proposed changes:

* Add event handlers for the error events emitted by `SuggestionsEventSource`
* Create custom onError callback to display the error on the editor, [using native editor notice support](https://developer.wordpress.org/block-editor/how-to-guides/notices/#notices-in-the-block-editor)

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* To test this, you need to force some errors on the backend; refer to the "How to force errors on the backend" section below to see how to do it
* With the backend changed to force an error return, go to the block editor
* Write some content on a paragraph
* Click the AI Assistant extension button and select some action:

<img width="628" alt="Screen Shot 2023-06-21 at 19 31 38" src="https://github.com/Automattic/jetpack/assets/6760046/6be1832b-6cbc-4729-bccd-a6fe8e6ea37b">

* Since you are forcing an error on the backend, you will see a notice on the top of the editor:

<img width="1180" alt="Screen Shot 2023-06-22 at 15 22 07" src="https://github.com/Automattic/jetpack/assets/6760046/dca5673f-2768-4a44-91e4-46a04b9a4be0">

* Change the `status` code returned on the backend error to the following values and repeat the test
   * 422
   * 429
   * 500
   * 503
* Each `status` code has it's own message on the frontend, so you will see different messages on each occasion

### How to force errors on the backend

* Sandbox your `public-api`
* If you are running this branch on JN, don't forget to point it to your sandbox
* Update the `jetpack-ai-query.php` file on your sandbox to add the following snippet:

```
return new \WP_Error( 'error_jetpack_ai', __( 'Generic error message.' ), [ 'status' => 503 ] );
```

* Add it as the first line of the function:

<img width="772" alt="Screen Shot 2023-06-21 at 19 29 33" src="https://github.com/Automattic/jetpack/assets/6760046/67a1c3ba-1f2f-400c-aa48-926d22cb1eb7">

* You will test changing the `status` value, so keep the file open and ready to edit
